### PR TITLE
Implement a PyPi version validator #1180

### LINKED ISSUE #1180: Implement a PyPi version validator
# Task Description

We need to implement version validators for PyPi.

For details on the PyPi spec for versions, check [here](https://www.python.org/dev/peps/pep-0440/).

For examples of existing version validators in Strongbox, check [here](https://strongbox.github.io/knowledge-base/artifact-coordinate-validators/artifact-coordinate-validators.html) and [here](https://strongbox.github.io/knowledge-base/artifact-coordinate-validators/artifact-coordinate-validators-maven.html).

Proper test cases would be required to illustrate that this works as expected.

# Task Relationships

This is a sub-task of #807 .

# Help

* [Our chat](https://chat.carlspring.org/)
* Points of contact:
  * @carlspring
  * @sbespalov
  * @fuss86

### PR DESCRIPTION
**Pull request by sainalshah**
Implement PyPi Canonical version validator following PEP 440

# Pull Request Description

This pull request closes #1180

# Acceptance Test

* [x] Building the code with `mvn clean install -Dintegration.tests` still works.
* [x] Running `mvn spring-boot:run` in the `strongbox-web-core` still starts up the application correctly.
* [x] Building the code and running the `strongbox-distribution` from a `zip` or `tar.gz` still works.
* [x] The tests in the [`strongbox-web-integration-tests`](https://github.com/strongbox/strongbox-web-integration-tests/) still run properly.

# Questions

* Does this pull request break backward compatibility? 
  * [ ] Yes
  * [x] No

* Does this pull request require other pull requests to be merged first? 
  * [ ] Yes, please see #...
  * [x] No

* Does this require an update of the documentation?
  * [ ] Yes, please see strongbox/strongbox-docs#{PR_NUMBER}
  * [x] No

### CODE REVIEW

**Comment by sainalshah** [comment]
I'm not entirely sure if this changes are enough, suggestions are welcome :)

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-pypi-layout-provider/src/main/java/org/carlspring/strongbox/storage/validation/version/PypiCanonicalVersionValidator.java [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+    public static final Pattern DEVELOPMENTAL_RELEASE_PATTERN = Pattern.compile(
+            PypiCanonicalVersionValidator.DEVELOPMENTAL_RELEASE_FORMAT);
+    public static final Pattern LOCAL_VERSION_IDENTIFIER_PATTERN = Pattern.compile(
+            PypiCanonicalVersionValidator.LOCAL_VERSION_IDENTIFIER_FORMAT);
+    public static final Pattern CANONICAL_PATTERN = Pattern.compile(
+            PypiCanonicalVersionValidator.CANONICAL_VERSION_FORMAT);
+
+    @Inject
+    private ArtifactCoordinatesValidatorRegistry artifactCoordinatesValidatorRegistry;
+    private static final Logger logger = LoggerFactory.getLogger(ArtifactCoordinatesValidatorRegistry.class);
```
Comment (ptirador) [inline strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-pypi-layout-provider/src/main/java/org/carlspring/strongbox/storage/validation/version/PypiCanonicalVersionValidator.java]: Please, use the current class here to get the logger.

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-pypi-layout-provider/src/main/java/org/carlspring/strongbox/storage/validation/version/PypiCanonicalVersionValidator.java [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+    }
+
+    @PostConstruct
+    @Override
+    public void register()
+    {
+        artifactCoordinatesValidatorRegistry.addProvider(ALIAS, this);
+
+        logger.info("Registered artifact coordinates validator '" + getClass().getCanonicalName() + "'" +
+                    " with alias '" + ALIAS + "'.");
```
Comment (ptirador) [inline strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-pypi-layout-provider/src/main/java/org/carlspring/strongbox/storage/validation/version/PypiCanonicalVersionValidator.java]: Please, use built-in parameters with logger methods.
e.g.:  `logger.info("Nothing found in db for A with a : {} and b : {}", a ,b);`

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-pypi-layout-provider/src/main/java/org/carlspring/strongbox/storage/validation/version/PypiCanonicalVersionValidator.java [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+            PypiCanonicalVersionValidator.POST_RELEASE_FORMAT);
+    public static final Pattern DEVELOPMENTAL_RELEASE_PATTERN = Pattern.compile(
+            PypiCanonicalVersionValidator.DEVELOPMENTAL_RELEASE_FORMAT);
+    public static final Pattern LOCAL_VERSION_IDENTIFIER_PATTERN = Pattern.compile(
+            PypiCanonicalVersionValidator.LOCAL_VERSION_IDENTIFIER_FORMAT);
+    public static final Pattern CANONICAL_PATTERN = Pattern.compile(
+            PypiCanonicalVersionValidator.CANONICAL_VERSION_FORMAT);
+
+    @Inject
+    private ArtifactCoordinatesValidatorRegistry artifactCoordinatesValidatorRegistry;
```
Comment (ptirador) [inline strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-pypi-layout-provider/src/main/java/org/carlspring/strongbox/storage/validation/version/PypiCanonicalVersionValidator.java]: Please, avoid field injection and try to use constructor injection:
```
private final ArtifactCoordinatesValidatorRegistry artifactCoordinatesValidatorRegistry;

public PypiCanonicalVersionValidator(final ArtifactCoordinatesValidatorRegistry artifactCoordinatesValidatorRegistry)
{
    this.artifactCoordinatesValidatorRegistry = artifactCoordinatesValidatorRegistry;
}
```

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-pypi-layout-provider/src/main/java/org/carlspring/strongbox/storage/validation/version/PypiCanonicalVersionValidator.java [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+
+    public static final Pattern PRE_RELEASE_PATTERN = Pattern.compile(PypiCanonicalVersionValidator.PRE_RELEASE_FORMAT);
+    public static final Pattern POST_RELEASE_PATTERN = Pattern.compile(
+            PypiCanonicalVersionValidator.POST_RELEASE_FORMAT);
+    public static final Pattern DEVELOPMENTAL_RELEASE_PATTERN = Pattern.compile(
+            PypiCanonicalVersionValidator.DEVELOPMENTAL_RELEASE_FORMAT);
+    public static final Pattern LOCAL_VERSION_IDENTIFIER_PATTERN = Pattern.compile(
+            PypiCanonicalVersionValidator.LOCAL_VERSION_IDENTIFIER_FORMAT);
+    public static final Pattern CANONICAL_PATTERN = Pattern.compile(
+            PypiCanonicalVersionValidator.CANONICAL_VERSION_FORMAT);
```
Comment (ptirador) [inline strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-pypi-layout-provider/src/main/java/org/carlspring/strongbox/storage/validation/version/PypiCanonicalVersionValidator.java]: Why are these members `public`? I think they can be `private`.

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-pypi-layout-provider/src/test/java/org/carlspring/strongbox/util/PypiCanonicalVersionValidatorTest.java [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+ */
+@SpringBootTest
+@ActiveProfiles(profiles = "test")
+@Execution(CONCURRENT)
+@ContextConfiguration(classes = PypiLayoutProviderTestConfig.class)
+public class PypiCanonicalVersionValidatorTest
+{
+
+    @Autowired
+    PypiVersionValidator pypiCanonicalVersionValidator;
```
Comment (ptirador) [inline strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-pypi-layout-provider/src/test/java/org/carlspring/strongbox/util/PypiCanonicalVersionValidatorTest.java]: Please, add `private` modifier here.

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-pypi-layout-provider/src/main/java/org/carlspring/strongbox/storage/validation/version/PypiCanonicalVersionValidator.java [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+    {
+        artifactCoordinatesValidatorRegistry.addProvider(ALIAS, this);
+
+        logger.info("Registered artifact coordinates validator '" + getClass().getCanonicalName() + "'" +
+                    " with alias '" + ALIAS + "'.");
+    }
+
+    public boolean isPreRelease(String version)
+    {
+        return PypiCanonicalVersionValidator.PRE_RELEASE_PATTERN.matcher(version).matches();
```
Comment (ptirador) [inline strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-pypi-layout-provider/src/main/java/org/carlspring/strongbox/storage/validation/version/PypiCanonicalVersionValidator.java]: Tip: You can use here (and in the next methods) a `BiPredicate`.
```
private BiPredicate<Pattern, String> versionMatches = (pattern, version) -> pattern.matcher(version).matches();

[...]

public boolean isPreRelease(String version)
{
     return versionMatches.test(PRE_RELEASE_PATTERN, version);
}

**Comment by sainalshah** [comment]
@ptirador Thanks for the inputs! Changes done.

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-pypi-layout-provider/src/main/java/org/carlspring/strongbox/storage/validation/version/PypiCanonicalVersionValidator.java [hunk 7]
```diff
@@ -123,4 +130,6 @@ private static String getCanonicalVersionFormat()
                DEVELOPMENTAL_RELEASE_FORMAT + "?" + LOCAL_VERSION_IDENTIFIER_FORMAT + "?)$";
     }
 
+    private BiPredicate<Pattern, String> versionMatches = (pattern, version) -> pattern.matcher(version).matches();
```
Comment (ptirador) [inline strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-pypi-layout-provider/src/main/java/org/carlspring/strongbox/storage/validation/version/PypiCanonicalVersionValidator.java]: Please, place this private member next to the others.

Comment (sainalshah) [reply strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-pypi-layout-provider/src/main/java/org/carlspring/strongbox/storage/validation/version/PypiCanonicalVersionValidator.java]: Done. Thanks.

**Comment by carlspring** [comment]
@sainalshah ,

Thanks for this great pull request! :)
Would you mind rebasing this and making sure that the tests pass, so that we could merge it?

@ptirador,

Thanks for reviewing!

**Comment by sainalshah** [comment]
@carlspring Sorry for the late reply. Tests are passing after the git rebase.

**Comment by carlspring** [comment]
Thank you very much!

**Comment by carlspring** [comment]
@anki2189 , @steve-todorov ,

What do we need to do to get the `strongbox-web-integration-tests` to pass?

**Comment by steve-todorov** [comment]
@carlspring You ran this PR against the `s-w-i-t#master` from 2 days ago, which did not include strongbox/strongbox-web-integration-tests#59 (which actually has the `pypi` s-w-i-t module). 

@anki2189 are there any `s-w-i-t` tests which need to be ran against this PR or is it independent? We had to temporarily disable the `pypi` s-w-i-t module.

**Comment by anki2189** [comment]
> @carlspring You ran this PR against the `s-w-i-t#master` from 2 days ago, which did not include [strongbox/strongbox-web-integration-tests#59](https://github.com/strongbox/strongbox-web-integration-tests/pull/59) at the time (and it actually has the `pypi` s-w-i-t module).
> 
> @anki2189 are there any `s-w-i-t` tests which need to be ran against this PR or is it independent? We had to temporarily disable the `pypi` s-w-i-t module.

I don't think so.

**Comment by sainalshah** [comment]
@carlspring rebased the branch against latest changes

**Comment by steve-todorov** [comment]
Thank you very much for your work on this @sainalshah! Sorry it has taken forever to merge it.

**Pull request merged**